### PR TITLE
Deprecate implicit tactic solving.

### PR DESCRIPTION
**Pull request by ppedrot**
That's about it. Does it deserve a CHANGES entry or is that going to be automatic?

### CODE REVIEW

**Review by SkySkimmer — DISMISSED** [review]
Doc needs to mention the deprecation.

Why is this deprecated anyway?

File: plugins/ltac/extratactics.ml4
Comment (herbelin) [inline plugins/ltac/extratactics.ml4]: The current convention (even if fragile) for errors is to add periods. For warnings, this is unregular. Maybe a good time to adopt a convention for warnings.

Comment (herbelin) [reply plugins/ltac/extratactics.ml4]: Would there be a way to shortly explain what to use instead (relying on a notation of the style "__ := ltac:(auto)")?

**Review by herbelin — COMMENTED** [review]
See comments.

**Comment by ppedrot** [comment]
I've updated the doc. It now recommends using typeclasses or tactics-in-terms as a replacement. AFAICT, the implicit mechanism was introduced in 8.1 and typeclasses in 8.2. Most simple uses fall in the scope of typeclasses, and fancy uses can be implemented with tactics in terms, e.g. using a notation for elegance as mentioned by @herbelin.

File: doc/sphinx/proof-engine/tactics.rst
Comment (gares) [inline doc/sphinx/proof-engine/tactics.rst]: I once heard you saying that tactics-in-terms was a mistake and now I see you recommend it.
What's happened?

(I personally like "the idea" tactics in terms, not quite their status today)

**Review by herbelin — DISMISSED** [review]
See comments.

File: doc/sphinx/proof-engine/tactics.rst
Comment (maximedenes) [reply doc/sphinx/proof-engine/tactics.rst]: Please use the `deprecated` directive: http://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated

Comment (Zimmi48) [reply doc/sphinx/proof-engine/tactics.rst]: Interesting! I'm going to try using this directive more in #7301.
Right now, it isn't used at all.

File: CHANGES
Comment (ppedrot) [reply CHANGES]: There are two commands, one to set it and one to unset it.

File: doc/sphinx/proof-engine/tactics.rst
Comment (ppedrot) [reply doc/sphinx/proof-engine/tactics.rst]: @herbelin I've added the crossrefs. I'm not too fond of the idea to add an replacement example though. It would make the documentation of that feature bigger than many other ones, while it is virtually unused in the wild. The few potential users have already switched to tactics-in-terms AFAICT.

Comment (herbelin) [reply doc/sphinx/proof-engine/tactics.rst]: > I'm not too fond of the idea to add an replacement example though. It would make the documentation of that feature bigger than many other ones, while it is virtually unused in the wild. The few potential users have already switched to tactics-in-terms AFAICT.

It stands up. OK with that.

File: doc/sphinx/proof-engine/tactics.rst
Comment (yurug) [reply doc/sphinx/proof-engine/tactics.rst]: Interested too by your answer to this question @ppedrot.

Comment (ppedrot) [reply doc/sphinx/proof-engine/tactics.rst]: Tactics-in-terms per se is OK, it's just that in the context of Ltac, it is a horrible source of hackish workarounds to the lack of expressivity of the language. So the feature is useful, but the uses are  often disgusting.

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]: This should be indented more `.. deprecated` should be at the same level as the preceding paragraph. In its current form, it is the cause of a Sphinx failure (see Travis).

Comment (ppedrot) [reply doc/sphinx/proof-engine/tactics.rst]: @Zimmi48 I've no idea about what I'm doing, let's hope it's working now.

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]: You did as I suggested and you've successfully managed to move this Sphinx warning a few lines down to the part that you didn't actually touched. Unfortunately, this `.. example::` and the inner `.. coqtop:: all` have to be similarly re-indented to be clearly on the same level as the preceding `.. deprecated::` and the paragraph above, inside the `.. cmd::` body. I am sorry that you are paying the price of a badly handled Sphinx migration :confused:

Comment (ppedrot) [reply doc/sphinx/proof-engine/tactics.rst]: Gaaah...

**Comment by Zimmi48** [comment]
For some reason it is still failing after the change I suggested. I'm sorry I don't know what to propose anymore. @maximedenes can you help?

File: doc/sphinx/proof-engine/tactics.rst
Comment (cpitclaudel) [inline doc/sphinx/proof-engine/tactics.rst]: Missing space after `` ` ``?

**Comment by herbelin** [comment]
Shall we wait for a review from @maximedenes? Seems otherwise that we could proceed and address (potential) users remarks or questions afterwards.

**Review by Zimmi48 — APPROVED** [review]
Formally validating the doc part as "acting" secondary doc maintainer. No need to wait for @maximedenes.

**Comment by ppedrot** [comment]
@herbelin ready to go.

**Comment by herbelin** [comment]
The 3 gitlab errors were apparently unrelated.

**Pull request merged**
